packages/next/src/client/components/action-async-storage.ts

```diff
@@ -0,0 +1,47 @@
+import type { AsyncLocalStorage } from 'async_hooks'
+
+export interface ActionStore {
+  readonly isAction: boolean
+}
+
+export type ActionAsyncStorage = AsyncLocalStorage<ActionStore>
+
+let createAsyncLocalStorage: () => ActionAsyncStorage
+if (process.env.NEXT_RUNTIME === 'edge') {
+  createAsyncLocalStorage = () => {
+    let store: ActionStore | undefined
+    return {
+      disable() {
+        throw new Error(
+          'Invariant: AsyncLocalStorage accessed in runtime where it is not available'
+        )
+      },
+      getStore() {
+        return store
+      },
+      async run<R>(s: ActionStore, fn: () => R): Promise<R> {
+        store = s
+        try {
+          return await fn()
+        } finally {
+          store = undefined
+        }
+      },
```

Comment:
Are edge functions multiplexed? The invocation here has some bugs around concurrent use of the ALS:
1. previous value isn't restored
2. the use of `await fn()` means the store is held for the async invocation of `fn`, when it should only be the _sync_ invocation of `fn` (async behavior inside `fn` will restore the store when necessary).
   - You're likely doing this to avoid the Promise patching and async transpilation that's necessary for a true polyfill, but the bug is still a problem if there's concurrent use.